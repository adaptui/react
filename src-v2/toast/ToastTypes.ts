export interface DefaultToast {
  id: string;
  visible: boolean;
  reverseOrder: boolean;
  animationDuration: number;
}

export type TimerToast = DefaultToast & {
  createdAt: number;
  pauseDuration: number;
  pausedAt: number | null;
  autoDismiss: boolean;
  dismissDuration: number;
};

export interface State<T> {
  toasts: T[];
}

export enum ActionType {
  ADD_TOAST,
  UPSERT_TOAST,
  UPDATE_TOAST,
  UPDATE_FIELD_TOAST,
  UPDATE_ALL_TOAST,
  DISMISS_TOAST,
  REMOVE_TOAST,
}

export type Action<T> =
  | {
      type: ActionType.ADD_TOAST;
      toast: T;
      maxToasts?: number;
    }
  | {
      type: ActionType.UPSERT_TOAST;
      toast: T;
    }
  | {
      type: ActionType.UPDATE_TOAST;
      toast: Partial<T>;
    }
  | {
      type: ActionType.UPDATE_FIELD_TOAST;
      field: keyof T;
      fieldValue: any;
      toast: Partial<T>;
    }
  | {
      type: ActionType.UPDATE_ALL_TOAST;
      toast: Partial<T>;
    }
  | {
      type: ActionType.DISMISS_TOAST;
      toastId?: string;
    }
  | {
      type: ActionType.REMOVE_TOAST;
      toastId?: string;
    };
