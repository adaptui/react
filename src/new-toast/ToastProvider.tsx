import * as React from "react";
import { createContext } from "@chakra-ui/utils";

import { Toast, useToastState, State, Action } from "./index";

export type ToastStore = State<Toast> & {
  dispatch: React.Dispatch<Action<Toast>>;
};

const [ToastStoreProvider, useToastStore] = createContext<ToastStore>({
  errorMessage: "useToastStore must be used within ToastProvider",
  name: "ToastsState",
  strict: false,
});

export { useToastStore };

export const ToastProvider = (props: any) => {
  const { children } = props;
  const store = useToastState<Toast>();

  return <ToastStoreProvider value={store}>{children}</ToastStoreProvider>;
};
