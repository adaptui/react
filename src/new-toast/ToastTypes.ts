export type Renderable = JSX.Element | string | number | null;

export type ValueFunction<TValue, TArg> = (arg: TArg) => TValue;
export type ValueOrFunction<TValue, TArg> =
  | TValue
  | ValueFunction<TValue, TArg>;
export type Content = ValueOrFunction<Renderable, Toast>;

export type ToastTypes = "info" | "success" | "warning" | "error";

export type ToastPlacement =
  | "top-left"
  | "top-center"
  | "top-right"
  | "bottom-left"
  | "bottom-center"
  | "bottom-right";

export interface Toast {
  createdAt: number;
  id: string;
  content: Content;
  visible: boolean;
  reverseOrder: boolean;
  pauseDuration: number;
  type: ToastTypes;
  placement: ToastPlacement;
  duration?: number;
  pausedAt?: number;
}

export type ToastOptions = Partial<
  Pick<Toast, "id" | "duration" | "type" | "placement" | "reverseOrder">
>;
