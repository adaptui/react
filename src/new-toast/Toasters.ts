import { ActionType, useToastStore, genId } from "./index";
import type { ToastOptions, Toast, Content } from "./index";

type ToastHandler = (content: Content, options?: ToastOptions) => string;

const createToast = (content: Content, opts?: ToastOptions): Toast => ({
  createdAt: Date.now(),
  visible: false,
  pauseDuration: 0,
  reverseOrder: false,
  type: "info",
  placement: "top-right",
  duration: Infinity,
  content,
  ...opts,
  id: opts?.id || genId(),
});

const useAddToast = (): ToastHandler => {
  const { dispatch } = useToastStore();

  return (content, options) => {
    const toast = createToast(content, options);
    dispatch({
      type: ActionType.UPSERT_TOAST,
      toast: { ...toast, visible: true },
    });

    return toast.id;
  };
};

const useShowToast = (): ToastHandler => {
  const { dispatch } = useToastStore();

  return (content, options) => {
    const toast = createToast(content, options);
    dispatch({ type: ActionType.UPSERT_TOAST, toast });

    setTimeout(() => {
      dispatch({
        type: ActionType.UPSERT_TOAST,
        toast: { ...toast, visible: true },
      });
    }, 0);

    return toast.id;
  };
};

const useDismissToast = () => {
  const { dispatch } = useToastStore();

  return (toastId?: string) => {
    dispatch({
      type: ActionType.DISMISS_TOAST,
      toastId,
    });

    setTimeout(() => {
      dispatch({
        type: ActionType.REMOVE_TOAST,
        toastId,
      });
    }, 1000);
  };
};

const useRemoveToast = () => {
  const { dispatch } = useToastStore();

  return (toastId?: string) => {
    dispatch({
      type: ActionType.REMOVE_TOAST,
      toastId,
    });
  };
};

// Toast Triggers
export const useToasters = () => {
  const addToast = useAddToast();
  const showToast = useShowToast();
  const dismissToast = useDismissToast();
  const removeToast = useRemoveToast();

  return { addToast, showToast, dismissToast, removeToast };
};
