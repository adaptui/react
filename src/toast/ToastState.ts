import * as React from "react";

interface Toast {
  id: string;
  visible: boolean;
  pauseDuration: number;
  reverseOrder?: boolean;
}

export interface State<T> {
  toasts: T[];
}

const TOAST_LIMIT = 20;

export enum ActionType {
  ADD_TOAST,
  UPDATE_TOAST,
  UPSERT_TOAST,
  DISMISS_TOAST,
  REMOVE_TOAST,
}

export type Action<T> =
  | {
      type: ActionType.ADD_TOAST;
      toast: T;
    }
  | {
      type: ActionType.UPSERT_TOAST;
      toast: T;
    }
  | {
      type: ActionType.UPDATE_TOAST;
      toast: Partial<T>;
    }
  | {
      type: ActionType.DISMISS_TOAST;
      toastId?: string;
    }
  | {
      type: ActionType.REMOVE_TOAST;
      toastId?: string;
    };

export const toastReducer = <T extends Toast>(
  state: State<T>,
  action: Action<T>,
): State<T> => {
  switch (action.type) {
    case ActionType.ADD_TOAST:
      return {
        ...state,
        toasts: action.toast.reverseOrder
          ? [...state.toasts, action.toast].slice(0, TOAST_LIMIT)
          : [action.toast, ...state.toasts].slice(0, TOAST_LIMIT),
      };

    case ActionType.UPDATE_TOAST:
      return {
        ...state,
        toasts: state.toasts.map(t =>
          t.id === action.toast.id ? { ...t, ...action.toast } : t,
        ),
      };

    case ActionType.UPSERT_TOAST:
      const { toast } = action;
      return state.toasts.find(t => t.id === toast.id)
        ? toastReducer(state, { type: ActionType.UPDATE_TOAST, toast })
        : toastReducer(state, { type: ActionType.ADD_TOAST, toast });

    case ActionType.DISMISS_TOAST:
      return {
        ...state,
        toasts: state.toasts.map(t =>
          t.id === action.toastId || action.toastId === undefined
            ? {
                ...t,
                visible: false,
              }
            : t,
        ),
      };

    case ActionType.REMOVE_TOAST:
      if (action.toastId === undefined) {
        return {
          ...state,
          toasts: [],
        };
      }

      return {
        ...state,
        toasts: state.toasts.filter(t => t.id !== action.toastId),
      };
  }
};

const initialState = { toasts: [] };

export const useToastState = <T extends Toast>(): StateReturnType<T> => {
  const [state, dispatch] = React.useReducer<
    React.Reducer<State<T>, Action<T>>
  >(toastReducer, initialState);

  return { toasts: state.toasts, dispatch };
};

export interface StateReturnType<T> {
  toasts: T[];
  dispatch: React.Dispatch<Action<T>>;
}
