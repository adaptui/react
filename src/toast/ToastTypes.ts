export type Renderable = JSX.Element | string | number | null;

export type ValueFunction<Value, Arg> = (arg: Arg) => Value;

export type ValueOrFunction<Value, Arg> = Value | ValueFunction<Value, Arg>;

export type Content = ValueOrFunction<Renderable, Toast>;

export type ToastTypes = "info" | "success" | "warning" | "error";

export type ToastPlacement =
  | "top-left"
  | "top-center"
  | "top-right"
  | "bottom-left"
  | "bottom-center"
  | "bottom-right";

export interface Toast {
  createdAt: number;
  id: string;
  visible: boolean;
  pauseDuration: number;
  content: Content;
  pausedAt?: number;
  type?: ToastTypes;
  placement?: ToastPlacement;
  duration?: number;
  reverseOrder?: boolean;
}

export type ToastOptions = Partial<
  Pick<Toast, "id" | "duration" | "type" | "placement" | "reverseOrder">
>;

export type DefaultToastOptions = Omit<ToastOptions, "id">;
