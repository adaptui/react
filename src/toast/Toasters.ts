import {
  genId,
  Toast,
  Content,
  ActionType,
  useToastStore,
  CreateToastOptions,
} from "./index";

export type ToastHandler = (
  content: Content,
  options?: CreateToastOptions,
) => string;

export type CreateToastHandler = (
  content: Content,
  options?: CreateToastOptions,
) => Toast;

export const useCreateToast = (): CreateToastHandler => {
  const { defaultOptions } = useToastStore();

  return (content, opts) => ({
    createdAt: Date.now(),
    visible: false,
    pausedAt: null,
    pauseDuration: 0,
    content,
    ...defaultOptions,
    ...opts,
    id: opts?.id || genId(),
  });
};

export const useAddToast = (): ToastHandler => {
  const { dispatch } = useToastStore();
  const createToast = useCreateToast();

  return (content, options) => {
    const toast = createToast(content, options);

    dispatch({
      type: ActionType.ADD_TOAST,
      toast: { ...toast, visible: true },
    });

    return toast.id;
  };
};

export const useShowToast = (): ToastHandler => {
  const { dispatch } = useToastStore();
  const createToast = useCreateToast();

  return (content, options) => {
    const toast = createToast(content, options);

    dispatch({ type: ActionType.ADD_TOAST, toast });

    setTimeout(() => {
      dispatch({
        type: ActionType.UPDATE_TOAST,
        toast: { ...toast, visible: true },
      });
    }, 0);

    return toast.id;
  };
};

export const useUpsertToast = (): ToastHandler => {
  const { dispatch } = useToastStore();
  const createToast = useCreateToast();

  return (content, options) => {
    const toast = createToast(content, options);

    dispatch({ type: ActionType.UPSERT_TOAST, toast });

    return toast.id;
  };
};

export const useUpdateToast = () => {
  const { dispatch } = useToastStore();

  return (toastId: string, toast: Partial<Toast>) => {
    dispatch({
      type: ActionType.UPDATE_TOAST,
      toast: { ...toast, id: toastId },
    });
  };
};

export const useDismissToast = () => {
  const { dispatch, defaultOptions } = useToastStore();

  return (toastId?: string) => {
    const unmountDuration = defaultOptions.animationDuration;

    dispatch({ type: ActionType.DISMISS_TOAST, toastId });

    setTimeout(() => {
      dispatch({ type: ActionType.REMOVE_TOAST, toastId });
    }, unmountDuration);
  };
};

export const useRemoveToast = () => {
  const { dispatch } = useToastStore();

  return (toastId?: string) =>
    dispatch({ type: ActionType.REMOVE_TOAST, toastId });
};

// Toast Triggers
export const useToasters = () => {
  const addToast = useAddToast();
  const showToast = useShowToast();
  const updateToast = useUpdateToast();
  const upsertToast = useUpsertToast();
  const dismissToast = useDismissToast();
  const removeToast = useRemoveToast();

  return {
    addToast,
    showToast,
    upsertToast,
    updateToast,
    dismissToast,
    removeToast,
  };
};
