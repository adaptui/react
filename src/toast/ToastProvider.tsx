import React from "react";
import ReactDOM from "react-dom";
import { canUseDOM } from "reakit-utils";
import { objectKeys } from "@chakra-ui/utils";

import { isFunction } from "../utils";
import { ToastController } from "./ToastController";
import { Toast, useToastState, ToastStateReturn } from "./ToastState";

const DEFAULT_TIMEOUT = 5000;
const PLACEMENTS = {
  "top-left": { top: 0, left: 0 },
  "top-center": { top: 0, left: "50%", transform: "translateX(-50%)" },
  "top-right": { top: 0, right: 0 },
  "bottom-left": { bottom: 0, left: 0 },
  "bottom-center": { bottom: 0, left: "50%", transform: "translateX(-50%)" },
  "bottom-right": { bottom: 0, right: 0 },
};

// let's infer the union types from the placement values instead of hardcoding them
export type Placements = keyof typeof PLACEMENTS;

interface ToastContextState extends ToastStateReturn {
  toastTypes: ToastTypes;
}

const ToastContext = React.createContext<ToastContextState | undefined>(
  undefined,
);
export const ToastContextProvider = ToastContext.Provider;

export function useToast() {
  const context = React.useContext(ToastContext);

  if (!context) {
    throw new Error(
      "The `useToasts` hook must be called from a descendent of the `ToastProvider`.",
    );
  }

  return context;
}

export type ToastTypes = Record<
  string,
  React.FC<
    Pick<Toast, "content" | "id" | "isVisible"> & {
      hideToast: ToastStateReturn["hideToast"];
    }
  >
>;

type ToastWrapper = (
  props: React.PropsWithChildren<{
    id: string;
    placement: Placements;
    isVisible?: boolean;
    index: number;
  }>,
) => any;

type ToastProviderProps = {
  /**
   * Specify types of toast in an object
   */
  toastTypes: ToastTypes;
  /**
   * Placement of the toast on the screen
   *
   * @default "bottom-center"
   */
  placement?: Placements;
  /**
   * If True the toast will automatically dismiss after the specified duration
   */
  autoDismiss?: boolean;
  /**
   * Timeout after the toast will be removed automatically if autoDismiss is true
   *
   * @default 5000
   */
  autoCloseTimeout?: number;
  /**
   * Duration of delay after the toast will be unmounted, so that animations can run
   *
   * @default 0
   */
  animationTimeout?: number;
  /**
   * Wrapper function to enhance the behaviour of ToastController
   */
  toastWrapper?: ToastWrapper;
};

export const ToastProvider: React.FC<ToastProviderProps> = ({
  children,
  toastTypes,
  placement: providerPlacement = "bottom-center",
  animationTimeout,
  toastWrapper: ToastWrapperComponent = ({ children }) => children,
  autoDismiss: providerAutoDismiss,
  autoCloseTimeout: providerTimeout = DEFAULT_TIMEOUT,
}) => {
  const state = useToastState({
    defaultPlacement: providerPlacement,
    animationTimeout,
  });
  const { sortedToasts, hideToast } = state;

  const Toasts = objectKeys(sortedToasts).map(placement => {
    const toastsList = sortedToasts[placement];

    return (
      <div
        key={placement}
        className={`toast__container toast__container--${placement}`}
        style={{
          position: "fixed",
          ...PLACEMENTS[placement],
        }}
      >
        {toastsList.map((toast, index) => {
          const { id, type, content, timeout, autoDismiss, isVisible } = toast;

          return (
            <ToastWrapperComponent
              key={id}
              id={id}
              index={index}
              isVisible={isVisible}
              placement={placement}
            >
              <ToastController
                id={id}
                onRequestRemove={hideToast}
                duration={timeout ?? providerTimeout}
                autoDismiss={autoDismiss ?? providerAutoDismiss}
              >
                {isFunction(content)
                  ? content({ id, isVisible, hideToast })
                  : toastTypes[type]?.({
                      id,
                      content,
                      isVisible,
                      hideToast,
                    }) || content}
              </ToastController>
            </ToastWrapperComponent>
          );
        })}
      </div>
    );
  });

  const portalTarget = canUseDOM ? document.body : null;

  return (
    <ToastContextProvider value={{ ...state, toastTypes }}>
      {children}
      {portalTarget ? ReactDOM.createPortal(Toasts, portalTarget) : Toasts}
    </ToastContextProvider>
  );
};
