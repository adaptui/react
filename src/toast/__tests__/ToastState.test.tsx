import React from "react";
import mockDate from "mockdate";
import { cleanup, renderHook } from "@testing-library/react-hooks";

import { Toast } from "../ToastTypes";
import { useCreateToast } from "../Toasters";
import { ToastProvider } from "../ToastProvider";
import { State, ActionType, Action, toastReducer } from "../ToastState";

beforeEach(() => {
  mockDate.set(new Date(1614683631209));
});
afterEach(cleanup);

const defaultState = { toasts: [] };

const renderToastHook = (initialState: State<Toast> = defaultState) => {
  return renderHook(
    () => {
      const values = React.useReducer<
        React.Reducer<State<Toast>, Action<Toast>>
      >(toastReducer, initialState);
      return {
        state: values[0],
        dispatch: values[1],
      };
    },
    { wrapper: ToastProvider },
  );
};

describe("ToastState reducer", () => {
  expect.assertions(1);
  it("should have initial state", () => {
    const { result } = renderToastHook();

    expect(result.current.state).toEqual({ toasts: [] });
  });

  it("should add a new toast", async () => {
    const { result } = renderToastHook();

    const {
      result: { current: toast },
    } = renderHook(() => useCreateToast(), { wrapper: ToastProvider });

    result.current.dispatch({
      type: ActionType.ADD_TOAST,
      toast: toast("Hello world", {
        autoDismiss: false,
        dismissDuration: Infinity,
      }),
    });

    expect(result.current.state).toEqual({
      toasts: [
        {
          animationDuration: 0,
          autoDismiss: false,
          content: "Hello world",
          createdAt: 1614683631209,
          dismissDuration: Infinity,
          id: "1",
          pauseDuration: 0,
          pausedAt: null,
          placement: "bottom-center",
          reverseOrder: false,
          type: "info",
          visible: false,
        },
      ],
    });
  });

  it("should add a new toast in reverse order", async () => {
    const initialState: State<Toast> = {
      toasts: [
        {
          animationDuration: 0,
          autoDismiss: false,
          content: "Hello world",
          createdAt: 1614683631209,
          dismissDuration: Infinity,
          id: "a",
          pauseDuration: 0,
          pausedAt: null,
          placement: "bottom-center",
          reverseOrder: false,
          type: "info",
          visible: false,
        },
        {
          animationDuration: 0,
          autoDismiss: false,
          content: "Hello world",
          createdAt: 1614683631209,
          dismissDuration: Infinity,
          id: "b",
          pauseDuration: 0,
          pausedAt: null,
          placement: "bottom-center",
          reverseOrder: false,
          type: "info",
          visible: false,
        },
      ],
    };
    const { result } = renderToastHook(initialState);

    const {
      result: { current: toast },
    } = renderHook(() => useCreateToast(), { wrapper: ToastProvider });

    result.current.dispatch({
      type: ActionType.ADD_TOAST,
      toast: toast("Helo world", {
        reverseOrder: true,
        autoDismiss: false,
        dismissDuration: Infinity,
      }),
    });

    expect(result.current.state).toMatchSnapshot();
  });

  it("should remove a toast", async () => {
    const initialState: State<Toast> = {
      toasts: [
        {
          animationDuration: 0,
          autoDismiss: false,
          content: "Hello world",
          createdAt: 1614683631209,
          dismissDuration: Infinity,
          id: "1",
          pauseDuration: 0,
          pausedAt: null,
          placement: "bottom-center",
          reverseOrder: false,
          type: "info",
          visible: false,
        },
      ],
    };

    const { result } = renderToastHook(initialState);

    expect(result.current.state.toasts.length).toEqual(1);

    result.current.dispatch({
      type: ActionType.REMOVE_TOAST,
      toastId: "1",
    });

    expect(result.current.state.toasts.length).toEqual(0);
  });

  it("should update a toast", async () => {
    const initialState: State<Toast> = {
      toasts: [
        {
          animationDuration: 0,
          autoDismiss: false,
          content: "Hello world",
          createdAt: 1614683631209,
          dismissDuration: Infinity,
          id: "1",
          pauseDuration: 0,
          pausedAt: null,
          placement: "bottom-center",
          reverseOrder: false,
          type: "info",
          visible: false,
        },
      ],
    };

    const { result } = renderToastHook(initialState);

    expect(result.current.state.toasts[0].content).toEqual("Hello world");

    result.current.dispatch({
      type: ActionType.UPDATE_TOAST,
      toast: { id: "1", content: "Updated!" },
    });

    expect(result.current.state.toasts[0].content).toEqual("Updated!");
  });

  it("should dissmiss a toast", async () => {
    const initialState: State<Toast> = {
      toasts: [
        {
          animationDuration: 0,
          autoDismiss: false,
          content: "Hello world",
          createdAt: 1614683631209,
          dismissDuration: Infinity,
          id: "1",
          pauseDuration: 0,
          pausedAt: null,
          placement: "bottom-center",
          reverseOrder: false,
          type: "info",
          visible: true,
        },
      ],
    };

    const { result } = renderToastHook(initialState);

    expect(result.current.state.toasts[0].visible).toBe(true);

    result.current.dispatch({
      type: ActionType.DISMISS_TOAST,
      toastId: "1",
    });

    expect(result.current.state.toasts[0].visible).toBe(false);
  });

  it("should upsert a toast", async () => {
    const initialState: State<Toast> = {
      toasts: [
        {
          animationDuration: 0,
          autoDismiss: false,
          content: "Hello world",
          createdAt: 1614683631209,
          dismissDuration: Infinity,
          id: "1",
          pauseDuration: 0,
          pausedAt: null,
          placement: "bottom-center",
          reverseOrder: false,
          type: "info",
          visible: true,
        },
      ],
    };

    const { result } = renderToastHook(initialState);

    const {
      result: { current: toast },
    } = renderHook(() => useCreateToast(), { wrapper: ToastProvider });

    result.current.dispatch({
      type: ActionType.UPSERT_TOAST,
      toast: toast("Second toast"),
    });

    expect(result.current.state.toasts.length).toBe(2);

    result.current.dispatch({
      type: ActionType.UPSERT_TOAST,
      toast: toast("Upserted toast", { id: "1" }),
    });

    expect(result.current.state.toasts.length).toBe(2);
    expect(result.current.state.toasts[1].content).toBe("Upserted toast");
  });
});
