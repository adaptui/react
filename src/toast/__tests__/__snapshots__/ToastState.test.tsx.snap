// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastState reducer should add a new toast in reverse order 1`] = `
Object {
  "toasts": Array [
    Object {
      "animationDuration": 0,
      "autoDismiss": false,
      "content": "Hello world",
      "createdAt": 1614683631209,
      "dismissDuration": Infinity,
      "id": "a",
      "pauseDuration": 0,
      "pausedAt": null,
      "placement": "bottom-center",
      "reverseOrder": false,
      "type": "info",
      "visible": false,
    },
    Object {
      "animationDuration": 0,
      "autoDismiss": false,
      "content": "Hello world",
      "createdAt": 1614683631209,
      "dismissDuration": Infinity,
      "id": "b",
      "pauseDuration": 0,
      "pausedAt": null,
      "placement": "bottom-center",
      "reverseOrder": false,
      "type": "info",
      "visible": false,
    },
    Object {
      "animationDuration": 0,
      "autoDismiss": false,
      "content": "Helo world",
      "createdAt": 1614683631209,
      "dismissDuration": Infinity,
      "id": "2",
      "pauseDuration": 0,
      "pausedAt": null,
      "placement": "bottom-center",
      "reverseOrder": true,
      "type": "info",
      "visible": false,
    },
  ],
}
`;
